### ISSUE TITLE
"Assertion `time == global_time' failed" when recording spidermonkey arm simulator

### ISSUE THREAD

**Issue by fitzgen** [mozilla/rr#1770]
I built rr from source on this commit:

```
commit cdb22ede3c3039950a77c77e05044f3fd01411ca
Author: Robert O'Callahan <robert@ocallahan.org>
Date:   Tue Aug 9 18:20:16 2016 +1200

    Awk string indices start at 1
```

I did a `ninja-build test` and all tests passed.

On Fedora with this kernel:

```
$ uname -a
Linux localhost.localdomain 4.6.5-300.fc24.x86_64 #1 SMP Thu Jul 28 01:10:12 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux
```

On this branch: https://github.com/fitzgen/gecko-dev/tree/rr-assertion-failure-test-case

Build the arm simulator like so:

```
$ ./js/src/devtools/automation/autospider.py --objdir $OBJDIR --build-only arm-sim
```

And then run like this:

```
bash-4.3$ ./js/src/tests/jstests.py $OBJDIR/dist/bin/js shell/futex.js -o --rr
## shell/futex.js: rc = -9, run time = 150.001963
Sleeping for 2 seconds
Waking the main thread now
Woke up as I should have in 2.011s
Main thread waiting on wakeup (2s)
Woke up as I should have in 2.011s
Starting outer wait
In the interrupt, starting inner wait with timeout 2s
Got the interrupt exception!
Script runs for too long, terminating.
TIMEOUT - shell/futex.js
[0|1|1|0] 100% ==========================================================>| 150.1s
TIMEOUTS
    shell/futex.js
FAIL
bash-4.3$ rr replay
rr: ../src/TraceStream.cc:543: void rr::TraceReader::read_generic(vector<uint8_t> &): Assertion `time == global_time' failed.
bash-4.3$ 
```

**Comment by Keno** [comment]
I've seen this happen when rr gets killed without being able to finish up writing the log.
Since it says `Script runs for too long, terminating.`, might rr have been killed during recording?

**Comment by rocallahan** [comment]
Good call. This line here also suggests rr was SIGKILLed.
> ## shell/futex.js: rc = -9, run time = 150.001963

I think we should, at minimum, track in the trace files whether rr terminated normally or not so we can provide a useful message here.

To optimally handle SIGKILL and also deal with some other future needs, I think the way to go would be to move trace compression and writing to a separate daemonized helper process that communicates with rr using a robust shared-mem protocol, so that if rr dies it can flush out the remaining trace writes. This could also be used as the basis for streaming traces to another machine or streaming them to a replay running in parallel with the recording. But that's quite a bit of work that I'm not eager to do right now.

**Comment by fitzgen** [comment]
Yeah this test is hanging and the jstests.py test runner is killing it after a time out. I suppose I can work around it by running the test manually without the test runner, but it would certainly be nice from a user's perspective if rr at least detected what happened and told me what was going on.
